41/create etl task

chriscasto opened the pull request:
# Pull Request Template

## Issue Overview

This PR addresses #issue

- [X] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This PR introduces the code for the ETL task that pulls listings from the Bloom endpoint and adds them into the database table created by #42.  It does not include the Dockerfile for that task

## How Can This Be Tested/Reviewed?

To run the unit tests, call `yarn junit` or `yarn test:cov` in the package root.  Test coverage output should look something like this:

<img width="630" alt="Screenshot 2023-04-20 at 11 01 49 AM" src="https://user-images.githubusercontent.com/23032560/233440990-3e44f698-c8ef-4c61-ab91-b43b9f2d5717.png">

To run the import process, you must first checkout the `31/create-combined-listings-endpoint` branch and run `yarn db:reseed` in the `backend/core` directory to create the database table.  Then switch back to this branch, go to `tasks/bloom-etl`, and run `yarn install` and `yarn build`.  Set the DATABASE_URL env var to your local postgres instance (probably `export DATABASE_URL=postgres://bloom-dev:bloom@localhost:5432/bloom`, and run `yarn import:run`.  You should get output that looks like this:

<img width="577" alt="image" src="https://user-images.githubusercontent.com/23032560/233445043-36dfb797-aea7-41a0-a6a7-9dc25451cae7.png">

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

chriscasto commented:
The Github linter appears to be having some trouble finding one of the packages.  I'm looking into it

lisunshiny commented on tasks/bloom-etl/index.ts:
```diff
@@ -0,0 +1,49 @@
+import { Runner, Extractor, Transformer, Loader, defaultMap } from "./src/etl"
+import { DbConfig, Jurisdiction, UrlInfo } from "./src/types"
+import { knex } from "knex"
```
@chriscasto  any idea what's happening with this?

lisunshiny replied:
OK @chriscasto I might know whats happening?

This is the github action. Yarn is only running in the top level dir but I believe you need to `yarn` from inside the tasks/bloom-etl dir to actualy install the packages into node_modules. I would either change yarn at the top level to install this too (I think that @ifranch did this with shared-services? not sure how though), or update the github action to yarn inside your dir.

```
jobs:
  run-linters:
    name: Run linters
    runs-on: ubuntu-latest

    steps:
      - name: Check out Git repository
        uses: actions/checkout@v2

      - name: Set up Node.js
        uses: actions/setup-node@v1
        with:
          node-version: 18

      # ESLint and Prettier must be in `package.json`
      - name: Install Node.js dependencies
        run: yarn install

      - name: Run linters
        run: yarn lint
```

chriscasto replied:
I'll try that out.  Thanks!

lisunshiny commented on tasks/bloom-etl/index.ts:
```diff
@@ -0,0 +1,49 @@
+
+// This is also unlikely to change during the lifetime of this task
+const url: UrlInfo = {
+  base: process.env.EXTERNAL_ENDPOINT_URL || "https://proxy.housingbayarea.org",
```
should we be consistent with the other services and put (all of) these into a .env(.template) file?

chriscasto replied:
Yeah, I can add that in no problem

chriscasto replied:
Let me check with Morgan on the best endpoint URL to use for non-prod.

chriscasto replied:
I removed the default value and put the dev endpoint in the .env.template file

lisunshiny commented on tasks/bloom-etl/test/etl/transformer.spec.ts:
```diff
@@ -0,0 +1,212 @@
+    const result = results[0]
+
+    // these are fields that directly map to object properties
+    expect(result).toHaveProperty("id", listing.id)
```
was looking for this test thanks!

lisunshiny reviewed (commented):
leaving some initial comments, will pull down branch to check out the yarn issue overall structure looks good though

chriscasto commented on tasks/bloom-etl/src/etl/extract/extractor.ts:
```diff
@@ -0,0 +1,72 @@
+
+  constructor(axios: Axios, urlInfo: UrlInfo, jurisdictions: Array<Jurisdiction>) {
+    super()
+    this.axios = axios
```
I used dependency injection here because the jest mocks weren't acting reliably and I needed a way to force the Extractor to use a specific instance of "axios", even if it wasn't axios proper.

chriscasto replied:
Added

chriscasto commented on tasks/bloom-etl/src/etl/transform/transformer.ts:
```diff
@@ -0,0 +1,46 @@
+      return obj[value as string]
+    }
+
+    if (type == "function") {
```
Map properties have two possible types: string or function.  If it's a string, the transformer looks for an API "listing" property with that name and gets the value from it.  If it's a function, it passes the "listing" into the function and uses the return value instead.  The most common case here is serializing a complex json object like buildingAddress into a string via `JSON.stringify`, but it's also possible to pull `listing.buildingAddress.city` into a top-level field this way.

lisunshiny replied:
can we add a comment here about this? I am still not super clear on why we need this but perhaps leaving a comment will help the next person figure it out

chriscasto commented:
@lisunshiny I changed the root eslintrc to ignore the etl dir and added a workflow to lint and test the ETL code.  Everything seems to be working, so please take another look when you get the chance.

Also looking for another reviewer
@ifranch 
@isaac-looker 
@phattya

lisunshiny commented on tasks/bloom-etl/index.ts:
```diff
@@ -0,0 +1,49 @@
+import axios from "axios"
+
+// No need to retrieve these dynamically since they never change, per Exygy
+const jurisdictions: Array<Jurisdiction> = [
```
Can we put the jurisdictions into the env variables as well ?if we are putting the api endpoint we should make it clear that these variables need to change too if that url changes -- OR we follow the pattern in Bloom and dynamically fetch these via /jurisdictions and use the jurisdiction names

chriscasto replied:
I thought about dynamically fetching them, but Morgan said the list of jurisdictions were static and would never change, so I hardcoded them to make it more clear and explicit about what jurisdictions are being used.  Honestly, I didn't even add an env var for the endpoint until the last minute, and that's mainly because I thought there might be a need to bypass the proxy in rare occasions.  Env vars don't do well with structured data and I like having the jurisdiction names for documentation and more useful log messages, so I think we want to keep all of that in.  

Keep in mind that this code will only have a lifetime of 6-12 months before it isn't needed any longer, so if this is really something that's worth investing the eng effort into, it might be worth looping back on the discussion, aligning on a direction, and addressing it in a new PR

chriscasto commented on tasks/bloom-etl/src/etl/transform/map.ts:
```diff
@@ -0,0 +1,60 @@
+
+  county: "countyCode",
+  city: (listing: Listing) => listing.buildingAddress?.city,
+  neighborhood: "neighborhood", // not available on view=base but needed for filtering
```
The `neighborhood` field is an existing filter that needs to be included for parity with the existing endpoint

chriscasto commented on tasks/bloom-etl/src/etl/transform/map.ts:
```diff
@@ -0,0 +1,60 @@
+export const defaultMap: RecordMap = {
+  id: "id",
+  assets: (listing: Listing) => jsonOrNull(listing.assets),
+  // household_size_min
```
There are a few fields I saw that we could use around household size and occupancy.  I wasn't sure which ones we would use, so originally added them all in.  I ended up removing these but left them in as a placeholder.  I can take them out though

chriscasto replied:
Removed

chriscasto commented on .eslintrc.js:
```diff
@@ -52,5 +52,6 @@ module.exports = {
     "migration/",
     "**/*.stories.tsx",
     "**/.eslintrc.js",
+    "tasks/bloom-etl"
```
The blended packages here are not the intended use case for yarn workspaces.  Each set of code should be treated as independent if we're following app dev best practices, and forcing tight coupling is detrimental to long-term maintainability.  Magic should be avoided, especially when it isn't documented.  Think about how difficult all the version upgrades have been because all the sites are sharing the same dependencies, many of which are not documented in their respective package.json files.  Or about how slow the build and test times are because they have to pull in dependencies for everything, even on a small local change, and how they are run as a single job rather than multiple in parallel.  I can run the linter and unit tests for this package in parallel, not just to other packages but to themselves, simply by not enforcing a dependency on all of the other code that, frankly, it shouldn't even know about, much less be dependent on.  When I run things locally, it takes single digit seconds to build, lint, and test rather than minutes, which is IMO a big QOL improvement from a dev perspective.

This code in particular is intended to be removed in a relatively short span of time, so keeping it separate makes it much easier to delete everything without any unintended side effects.  It also makes it much easier to pull it into a separate repo if needed at some point, as the only things that are needed are the files within this specific directory and the single workflow file that contains all the jobs needed to validate the code.  I don't think losing that to save a couple of lines in a couple of config files is worth it.

chriscasto commented on tasks/bloom-etl/.env.template:
```diff
@@ -0,0 +1,3 @@
+DATABASE_URL=postgres://bloom-dev:bloom@localhost:5432/bloom
+EXTERNAL_ENDPOINT_URL=https://api.bloom-endpoint.org
```
I do have a small nit with the "bloom" nomenclature, as that word primarily refers to an open source project rather than a specific implementation of that project.  It's not enough that I would ask it to be changed, but it's enough that I wouldn't want to propagate it.  I went with `EXTERNAL_ENDPOINT_URL` because it seemed descriptive in terms of what it represents and not overly restrictive (it could also be used to pull from Detroit or whatever), but I'm not necessarily stuck on that either.  

I do question whether there is value in providing a configurable path over something hardcoded given that there is no identifiable need to change that in the lifetime of this task.  I don't want to waste too many cycles solving problems that probably won't exist, especially if it would be a simple fix in another PR

lisunshiny replied:
in that case, can you please add a comment on this and the same template file in backend/core saying to update EXTERNAL_blah on the backend/core side and BLOOM_blah on this one? want to make sure that developers know that there is an implicit dependency on both

also noting that this dir is literally called `bloom-etl` re. concerns around bloom naming

chriscasto replied:
Yeah, that's a fair point about the dir name.  I picked it from the beginning to match and then just never changed it.  It's been bugging me, just not enough to do anything about it.  I'll probably change it later to something more descriptive like `listings-etl` so that it's more clear what is being imported

I'll be removing the BLOOM* vars on the backend side in a subsequent PR since they won't be needed anymore, so the dependency will be short-lived enough that it shouldn't be an issue for future developers.  The shared dependency on the public site will probably be around for a while though, so I can add a note there

chriscasto commented on tasks/bloom-etl/src/etl/extract/extractor.ts:
```diff
@@ -0,0 +1,72 @@
+  private constructEndpoint(id: string) {
+    // Construct endpoint urls that gets base info for all listings for each jurisdiction individually
+    // This enables us to maximize cache hits on the Bloom side
+    const url =
```
So just return the string directly rather than putting it in a var?  Yeah, I can do that.  The original implementation concatenated the first part then generated an array of URLs in a loop (hence the bifurcated concat), but that's definitely not needed anymore

chriscasto commented on tasks/bloom-etl/src/etl/extract/extractor.ts:
```diff
@@ -0,0 +1,72 @@
+      responses.forEach((result) => {
+        const respItems = result.response.data.items
+
+        this.log(`Retrieved ${respItems.length} listings from [${result.jurisdiction.name}]`)
```
I added this to the `Logger` class

chriscasto commented on tasks/bloom-etl/src/etl/extract/extractor.ts:
```diff
@@ -0,0 +1,72 @@
+          .get<Response>(endpoint)
+          .catch((error) => {
+            this.log(error)
+            throw new Error("Unexpected HTTP error")
```
I added the url and the jurisdiction name to the logs

chriscasto commented:
@lisunshiny I made additional changes per your comments.  Please resolve convos that no longer need attention or comment on ones that need further discussion
